File: scala/kindergarten-garden/src/main/scala/Garden.scala
class Garden(kids: List[String], plants: String) {

  lazy val sortedKids = kids.sorted

  val kidsToPlantsMap = {
    val plantRows = plants.split("\n").map(_.map(Plant.apply).grouped(2))
    val plantsForKids = plantRows(0).zip(plantRows(1)).map(r => r._1.toList ++ r._2.toList).toList

    sortedKids.zip(plantsForKids)
        .toMap
        .withDefaultValue(Nil)
  }

  def getPlants(kid: String) = kidsToPlantsMap(kid)
}

object Garden {
  private val defaultKids = List("Alice", "Bob", "Charlie", "David", "Eve", "Fred", "Ginny", "Harriet", "Ileana", "Joseph", "Kincaid", "Larry")

  def apply(kids: List[String], plants: String) = new Garden(kids, plants)
  def defaultGarden(plants: String) = new Garden(defaultKids, plants)
}

object Plant extends Enumeration {
  val Radishes = Value('R'.toInt, "Radishes")
  val Clover = Value('C'.toInt, "Clover")
  val Grass = Value('G'.toInt, "Grass")
  val Violets = Value('V'.toInt, "Violets")

  def apply(c: Char): Value = Plant.apply(c.toInt)
}

File: v1/haskell/kindergarten-garden/Garden.hs
module Garden (garden, defaultGarden, lookupPlants, Plant(..)) where

import Data.List (sort)
import Data.List.Split (chunksOf)
import Data.Map (Map, fromList, findWithDefault)

data Plant = Violets | Radishes | Clover | Grass
    deriving (Show, Eq)

plantsPerChildPerRow :: Int
plantsPerChildPerRow = 2

defaultChildren :: [String]
defaultChildren = ["Alice", "Bob", "Charlie", "David",
                   "Eve", "Fred", "Ginny", "Harriet",
                   "Ileana", "Joseph", "Kincaid", "Larry"] 

garden :: [String] -> String -> Map String [Plant]
garden children windowSills = fromList $ zipWith3 (\child plants1 plants2 -> (child, plants1 ++ plants2)) (sort children) row1 row2
    where rows = lines windowSills
          row1 = plantsInRow $ head rows
          row2 = plantsInRow $ last rows

defaultGarden :: String -> Map String [Plant]
defaultGarden = garden defaultChildren  

lookupPlants :: String -> Map String [Plant] -> [Plant]
lookupPlants child plantsMap = findWithDefault [] child plantsMap

plantFromCode :: Char -> Plant
plantFromCode 'V' = Violets
plantFromCode 'R' = Radishes
plantFromCode 'C' = Clover
plantFromCode 'G' = Grass
plantFromCode _   = error "Invalid plant code"

plantsInRow :: String -> [[Plant]]
plantsInRow = chunksOf plantsPerChildPerRow . map plantFromCode